deduper.c

```diff
@@ -0,0 +1,127 @@
+/*
+ * Copyright (c) 2021 by Farsight Security, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <stdbool.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+
+#include "deduper.h"
+
+struct chainlink;
```

Comment:
we use a different pattern in files like pdns.h:
```
struct pdns_tuple {
        struct pdns_json  obj;
...};
typedef struct pdns_tuple *pdns_tuple_t;
```

Reply:
in this case we're only noting that the struct type exists, which is enough to make a pointer for it. information hiding is used to prevent any code outside of deduper.c from accessing the inner fields of this structure. we aren't doing information hiding of this kind in pdns.[ch] because it's an integral part of the purpose of the program. with deduper.[ch] i feel there's a good chance it will be copied and used in other projects.